R/exploratoryfactoranalysis.R

```diff
@@ -196,13 +230,57 @@ ExploratoryFactorAnalysis <- function(jaspResults, dataset, options, ...) {
   modelContainer[["bartab"]] <- bartab
 
   if (!ready) return()
-  bar <- psych::cortest.bartlett(dataset)
+
+  # Modification here:
+  # if "basedOn = mixed", bartlett's test is calculated
+  # based on the polychoric/tetrachoric matrix.
+  if (options[["basedOn"]] == "mixed") {
+    polytetracor <- psych::mixedCor(dataset)
+    bar <- psych::cortest.bartlett(polytetracor$rho, n = nrow(dataset))
+  }
+  else {
+    bar <- psych::cortest.bartlett(dataset)
+  }
 
   bartab[["chisq"]] <- bar[["chisq"]]
   bartab[["df"]]    <- bar[["df"]]
   bartab[["pval"]]  <- bar[["p.value"]]
 }
 
+# Modification here:
+# Added Mardia's tests of multivariate normality for further probing of the
+# multivariate normality assumption.
+.efaMardia <- function(modelContainer, dataset, options, ready) {
+  if (!options[["martest"]] || !is.null(modelContainer[["martab"]])) return()
+
+  martab <- createJaspTable(gettext("Mardia's tests of Multivariate Normality"))
+  martab$dependOn("martest")
+  martab$addColumnInfo(name = "tests", title =  "", type = "number", format = "dp:3")
+  martab$addColumnInfo(name = "coefs", title =  gettext("Coefficient"), type = "number", format = "dp:3")
+  martab$addColumnInfo(name = "statistics", title =  gettext("Statistic"), type = "number", format = "dp:3")
+  martab$addColumnInfo(name = "dfs", title =  gettext("df"), type = "integer")
+  martab$addColumnInfo(name = "pval",  title = gettext("p"), type = "number", format = "dp:3;p:.001")
+  martab$position <- 0.5
+  modelContainer[["martab"]] <- martab
+
+  if (!ready) return()
+
+  mar <- psych::mardia(dataset, plot = F)
+
+  mardianames <- c("Skewness","Small Sample Skewness", "Kurtosis")
+
+  # dfs of the skewness coefficients are calculated via Mardia's formula (1970);
+  # package psych doesn't seem to provide them
+  k <- length(dataset)
+  mardiadfs <- (k*(k+1)*(k+2))/6
+
+  martab[["tests"]] <- mardianames
+  martab[["coefs"]] <- c(mar[["b1p"]], mar[["b1p"]], mar[["b2p"]])
+  martab[["statistics"]] <- c(mar[["skew"]], mar[["small.skew"]], mar[["kurtosis"]])
```

Comment:
I haven't seen an explicit rule - just a reference to its usage in "small samples" (I've seen it with either n < 20 or n < 30, although I cannot remember the references). On another note, I agree and will implement so that when n < 20 it switches to small sample skewness automatically.

Reply:
Well, if there is not specific rule then making a behind-the-scenes switch might not be the best idea. What if sample size is 25... I suppose the technically best solution would be two checkboxes, one for the normal Mardia test and one for the small sample Mardia test. But I think that would be a bit of an overkill for the assumption checks. So I think your initial solution might be the best for now. Sorry for the back and forth :)

Reply:
To be fair, all the other table names are exactly the style that @francotisocco used. Nevertheless, it would be good for any new code to follow the camelCase stuff. Some of the maintainers will have to go over the rest of the code some other time for fix the naming.

Reply:
Hi @AlexanderLyNL, thank you for pointing these things out - it's not nitpicky at all, and is definitely cleaner that way. @juliuspf I actually took the existing table names as reference so they ended up like that; I'll at least modify the `mardianames` so that it is clearer for future revisions